### PR TITLE
Arnej/shrink if needed

### PR DESCRIPTION
**Pull request by arnej27959**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@geirst please review

### CODE REVIEW

File: searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp [hunk 1]
```diff
@@ -287,8 +325,8 @@ HnswIndex::add_document(uint32_t docid)
     while (search_level >= 0) {
         // TODO: Rename to search_level?
         search_layer(input, _cfg.neighbors_to_explore_at_construction(), best_neighbors, search_level);
-        auto neighbors = select_neighbors(best_neighbors.peek(), max_links_for_level(search_level));
-        connect_new_node(docid, neighbors, search_level);
+        auto neighbors = select_neighbors(best_neighbors.peek(), _cfg.max_links_at_hierarchic_levels());
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp]: Shouldn't we still differentiate between max links for level 0 and hierarchical levels?

File: searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp [hunk 2]
```diff
@@ -287,8 +325,8 @@ HnswIndex::add_document(uint32_t docid)
...
+        connect_new_node(docid, neighbors.used, search_level);
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp]: I guess we can remove the TODO on the next line?

**Review by geirst — APPROVED** [review]
👍  Please have a look at the question and nitpick

File: searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp (hunk 1 above)
Comment (arnej27959) [reply searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp]: yes and no; we're supposed to select (up to ) M neighbors on all layers when inserting, but the "Mmax" used when shrinking is differentiatied (Mmax = M while Mmax0 = 2 * M)

File: searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp (hunk 2 above)
Comment (arnej27959) [reply searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp]: right!

File: searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp (hunk 1 above)
Comment (geirst) [reply searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp]: Ok, I see

**Pull request merged**